payara/monitoring-console
# FISH-527 Alert Centre 1.0

### Summary
Goal of this PR is to add "global" alert feedback to Payara InSight. 
This means a user can/will notice all alerts independent of the page currently shown.
This is approached in two ways, one more discreet and another one more right in your face.

**The discrete** is a alert status indicator in the footer that gives a summary of all alerts in form of numbers. Clicking on the indicator opens the _Alerts_ page where any ongoing alert (which these are) will be shown.

![mc_footer-alerts](https://user-images.githubusercontent.com/309438/96894315-e2063780-148b-11eb-8e01-80c81c799400.png)

**The persistent** is an alert status change popup which automatically pops up the moment an alert status change occurs.
It details the changes and shows the alert details.

![screenshot-localhost_8080-2020 10 22-17_26_41](https://user-images.githubusercontent.com/309438/96894625-0d892200-148c-11eb-8d4a-2ec8e99b6b5b.png)
 
The persistent one can be disabled by a new toggle settings in _Application_ settings tab _Alerts_; by default it is _On_.

![mc_settings_app_alerts-h](https://user-images.githubusercontent.com/309438/96895042-44f7ce80-148c-11eb-8b48-861929459077.png)

I also sneaked in a fix for the issue that scroll state would be lost due to update being applied to the component.
This now allows to go through lists of alerts without having the scroll jump to the top after 2 seconds (poll interval). :space_invader: 

### Design Decisions
Some decisions made when designing the new feature:

* **Alert Status Change popup is layered below other popups.** For example when being in the watches settings while an alert status change occurs the change is first encountered when the settings popup is closed. Reasoning: It would be too annoying otherwise
* **Alert Status Change popup is _not_ modal.** That means you can click on stuff in the background and the overlay is not greying out content outside of the dialog. Reasoning: It makes sense to be able to clearly see and even use what is on the page while seeing the popup on top. Technically this is still the same modal dialog just that overlay color is changed and it does not capture click events.
* **Transitions from _Healthy_ and _Normal_ are both shown as transitions from _Healthy_**. Reasoning: It is simpler and more clear to have less combinations (of state transitions) and for alerts it does not matter if the previous state was _Healthy_ or _Normal_, those are one group in that context.
* **Clicking _Show_ in the popup which navigates to the _Alerts_ page will also stop any ongoing page rotation.** Reasoning: As the user has the intention to now look at the alerts continuing rotation makes little sense
* **The alert indicator flashes whenever there are unacknowledged alerts.** Reasoning: This is usually when you are not aware of the alerts so attention should be drawn to the alerts, so they flash

### Technical Details
The alerts summary used to update the alert indicator has always been part of the response when polling data for the current page. This means there is no extra request and pages continue to only use a single request to poll data to update the page.
The payload has been extended slightly now also containing the serial numbers of the currently ongoing alerts which is considered a small set.

When alert details are shown as part of the status change popup this information potentially needs to be loaded from the server. But it might be included in the alerts attached to widget on the page. If so this information will be used and no extra requests are done. If the alert details needed are not present they are loaded by serial number creating one request per alert details. To prevent this from happening each time the page is updated in polling interval a check is added that can tell if the popup potentially already included in the page (from last update) already shows the correct popup. In that case the popup is just left untouched and no extra server requests are needed until the status actually changes again. If a change occurred can easily be identified by the change counter included in the alerts statistics. 

**TLDR;** Even with new feature we still just do single poll request, extra requests are avoided where possible and only occur once when needed until next actual status change.

### Testing
#### Setup
* Build InSight (all modules)
* copy `process` module to `glassfish/modules` folder
* deploy `webapp` module manually

(or do a full server build with updates references to InSight SNAPSHOT versions)

#### Testing Performed
* enable Payara Health Check for Machine Memory Usage
* configure warning threshold so that current usage will be triggered
* open InSight and wait for the popup informing about the status change, press ok or show
* disable the health check
* check in InSight another popup informs about the change, press ok or show
* re-enable the health check and check again for the popup
* repeat the above without pressing ok or show when popup occurs and check the popup updates with the changes to health check
* configure health check so critical threshold is exceeded and check status change is given by the popup
* disable popups in the _Application_ => tab _Alerts_ => _ property _Popups_ setting
* cause some alert status changes and check the alert indicator in the bottom right of the footer updates accordingly
* acknowledge an alert and check the alert indicator updates accordingly

Also I tested with an app including MP health checks which cause watches and alerts should the health check compute to status _DOWN_. 

* enable page rotation and repeat some of the testing to verify that the popup is shown even while rotating through the pages
* press _Show_ in the popup while pages rotate and check rotation stops after jumping to the _Alerts_ page

### Documentation
https://github.com/payara/Payara-Enterprise-Documentation/pull/68
